01_homework/hw2_task4.cpp

```diff
@@ -0,0 +1,183 @@
+#include <iostream>
+#include <vector>
+#include <thread>
+#include <chrono>
+#include <iomanip>
+#include <random>
+#include <mutex>
+#include <array>
+
+#undef LOG
+
+constexpr int dining_time = 10;
+constexpr int population = 5;
+constexpr int starvation_treshold = 50;
+
+std::mutex print_mutex;
+
+struct Fork
+{
+   std::mutex mutex;
+};
+
+struct Table
+{
+    bool food = true;
+
+    std::array<Fork, population> forks;
+    std::array<int, population> ownership;
+    std::array<int, population> starvation;
+
+    void print(bool lock, int id);
+    bool check_neighbour_stavation(int id);
+};
+
+#ifdef LOG
+void log_print(int id, std::string_view text)
+{
+   std::lock_guard<std::mutex> cout_lock(print_mutex);
+   std::cout << static_cast<char>(id+65) << text << std::endl;
+}
+#endif
+
+std::string print_fork(int value) {
+    if (value < 0) return "|  ";
+    else if (value == 0) return " - ";
+    else return "  |";
+}
+
+void Table::print(bool lock, int id) {
+    std::lock_guard<std::mutex> print_lock(print_mutex);
+
+    // Check border case
+    int tmp = id + 1;
+    if (tmp == population)
+        tmp = 0;
+
+    if (lock) {
+        ownership[id] = 1;
+        ownership[tmp] = -1;
+    } else {
+        ownership[id] = 0;
+        ownership[tmp] = 0;
+    }
+
+    std::stringstream buff;
+    const int ascii_value = 65;
+
+#ifdef LOG
+    buff << "[" << id << "," << lock << "]: ";
+#endif
+
+    for (int i = 0; i < population; i++) {
+        buff << print_fork(ownership[i]);
+        buff << static_cast<char>(i + ascii_value);
+#ifdef LOG
+        buff << check_neighbour_stavation(i) << "." << starvation[i];
+#endif
+    }
+    buff << print_fork(ownership[0]);
+    buff << '\n';
+
+    std::cout << buff.str();
+}
+
+bool Table::check_neighbour_stavation(int id) {
+    int neighbour_left = id - 1;
+    int neighbour_right = id + 1;
+
+    // Check border cases
+    if(neighbour_left < 0)
+        neighbour_left = population - 1;
+    if(neighbour_right == population)
+        neighbour_right = 0;
+
+    if(starvation[neighbour_left] - starvation[id] > starvation_treshold)
+        return false;
+    if(starvation[neighbour_right] - starvation[id] > starvation_treshold)
+        return false;
+
+    return true;
+}
+
+void eat(int id, Table &table, Fork &fork_left, Fork &fork_right, std::mt19937 &seed) {
+    int result = 0;
+
+    while(result != -1) {
+        if(table.check_neighbour_stavation(id)) {
+            result = std::try_lock(fork_left.mutex, fork_right.mutex);
```

Comment:
result chyba powinien być std::atomic.
U mnie są wyścigi, ale nie jestem pewien czy to nie coś jeszcze.
https://pastebin.com/uqFzW9xU

Reply:
wrzucilem wspodzielone zmienne w atomici, to bylo najprostsze rozwiazanie problemu. Smiga nawet przy 32 filozofach.